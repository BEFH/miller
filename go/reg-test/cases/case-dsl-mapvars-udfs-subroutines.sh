run_mlr --from $indir/abixy put '
  func f(m) {
    dump m;
    sum = 0;
    for (k, v in m) {
      sum += int(k)
    }
    return sum
  }
  @v[$i] = $a;
  $y = f(@v)
'

run_mlr --from $indir/abixy put '
  subr s(m) {
    dump m;
    sum = 0;
    for (k, v in m) {
      sum += int(k)
    }
    @sum = sum;
  }
  @v[$i] = $a;
  call s(@v);
  $y = @sum;
'

run_mlr --from $indir/abixy-het put    'func f(x) {return {"a":x,"b":x**2}}; map o = f($x); $* = o'
run_mlr --from $indir/abixy-het put -q 'func f(x) {return x**2}; var z = f($x); dump z'
run_mlr --from $indir/abixy-het put -q 'func f(x) {map m = {NR:x};return m}; z = f($y); dump z'

mlr_expect_fail --from $indir/abixy put '
  func f(int x): map {
    if (NR==2) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($x)
'

run_mlr --from $indir/abixy put '
  func f(int x): map {
    if (NR==200) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($i)
'

mlr_expect_fail --from $indir/abixy put '
  func f(int x): map {
    if (NR==200) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($x)
'

run_mlr --from $indir/abixy put '
  func f(int x): var {
    if (NR==2) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($i)
'

mlr_expect_fail --from $indir/abixy put '
  func f(int x): var {
    if (NR==2) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($x)
'

mlr_expect_fail --from $indir/abixy put '
  func f(x): int {
    # fall-through return value is absent-null
  }
  $y=f($x)
'

mlr_expect_fail --from $indir/abixy put '
  int a = 1;
  var b = a[2]; # cannot index localvar declared non-map
'

# This one is intended to particularly look at freeing, e.g.  with './reg-test/run --valgrind'.
run_mlr --oxtab --from $indir/abixy-het put '
 $* = mapdiff(
   mapsum($*, {"a": "newval"}),
   {"b": "nonesuch"},
 )
'
