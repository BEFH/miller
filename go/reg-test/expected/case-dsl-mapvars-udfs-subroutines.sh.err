
Miller: couldn't assign variable int x from value float 0.3467901443380824
Miller: couldn't assign variable int x from value float 0.3467901443380824
Miller: couldn't assign variable int x from value float 0.3467901443380824
Miller: couldn't assign variable int function return value from value absent (absent)
Miller: couldn't assign variable var b from value error (error)

