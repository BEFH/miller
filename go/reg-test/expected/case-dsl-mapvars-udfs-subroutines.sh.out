
mlr --from ./reg-test/input/abixy put 
  func f(m) {
    dump m;
    sum = 0;
    for (k, v in m) {
      sum += int(k)
    }
    return sum
  }
  @v[$i] = $a;
  $y = f(@v)

{
  "1": "pan"
}
a=pan,b=pan,i=1,x=0.3467901443380824,y=1
{
  "1": "pan",
  "2": "eks"
}
a=eks,b=pan,i=2,x=0.7586799647899636,y=3
{
  "1": "pan",
  "2": "eks",
  "3": "wye"
}
a=wye,b=wye,i=3,x=0.20460330576630303,y=6
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks"
}
a=eks,b=wye,i=4,x=0.38139939387114097,y=10
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye"
}
a=wye,b=pan,i=5,x=0.5732889198020006,y=15
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee"
}
a=zee,b=pan,i=6,x=0.5271261600918548,y=21
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks"
}
a=eks,b=zee,i=7,x=0.6117840605678454,y=28
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks",
  "8": "zee"
}
a=zee,b=wye,i=8,x=0.5985540091064224,y=36
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks",
  "8": "zee",
  "9": "hat"
}
a=hat,b=wye,i=9,x=0.03144187646093577,y=45
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks",
  "8": "zee",
  "9": "hat",
  "10": "pan"
}
a=pan,b=wye,i=10,x=0.5026260055412137,y=55

mlr --from ./reg-test/input/abixy put 
  subr s(m) {
    dump m;
    sum = 0;
    for (k, v in m) {
      sum += int(k)
    }
    @sum = sum;
  }
  @v[$i] = $a;
  call s(@v);
  $y = @sum;

{
  "1": "pan"
}
a=pan,b=pan,i=1,x=0.3467901443380824,y=1
{
  "1": "pan",
  "2": "eks"
}
a=eks,b=pan,i=2,x=0.7586799647899636,y=3
{
  "1": "pan",
  "2": "eks",
  "3": "wye"
}
a=wye,b=wye,i=3,x=0.20460330576630303,y=6
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks"
}
a=eks,b=wye,i=4,x=0.38139939387114097,y=10
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye"
}
a=wye,b=pan,i=5,x=0.5732889198020006,y=15
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee"
}
a=zee,b=pan,i=6,x=0.5271261600918548,y=21
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks"
}
a=eks,b=zee,i=7,x=0.6117840605678454,y=28
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks",
  "8": "zee"
}
a=zee,b=wye,i=8,x=0.5985540091064224,y=36
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks",
  "8": "zee",
  "9": "hat"
}
a=hat,b=wye,i=9,x=0.03144187646093577,y=45
{
  "1": "pan",
  "2": "eks",
  "3": "wye",
  "4": "eks",
  "5": "wye",
  "6": "zee",
  "7": "eks",
  "8": "zee",
  "9": "hat",
  "10": "pan"
}
a=pan,b=wye,i=10,x=0.5026260055412137,y=55

mlr --from ./reg-test/input/abixy-het put func f(x) {return {"a":x,"b":x**2}}; map o = f($x); $* = o
a=0.3467901443380824,b=0.12026340421002804
a=0.7586799647899636,b=0.5755952889737004
a=0.20460330576630303,b=0.04186251273049929
a=0.38139939387114097,b=0.14546549764527372
b=0
a=0.5271261600918548,b=0.2778619886531837
a=0.6117840605678454,b=0.37427973676488113
a=0.5985540091064224,b=0.3582669018173712
a=0.03144187646093577,b=0.0009885915953847467
a=0.5026260055412137,b=0.2526329014463162

mlr --from ./reg-test/input/abixy-het put -q func f(x) {return x**2}; var z = f($x); dump z
0.12026340421002804
0.5755952889737004
0.04186251273049929
0.14546549764527372
0
0.2778619886531837
0.37427973676488113
0.3582669018173712
0.0009885915953847467
0.2526329014463162

mlr --from ./reg-test/input/abixy-het put -q func f(x) {map m = {NR:x};return m}; z = f($y); dump z
{
  "1": 0.7268028627434533
}
{
  "2": 0.5221511083334797
}
{
  "3": 0.33831852551664776
}
{
  "4": 0.13418874328430463
}
{
  "5": 0.8636244699032729
}
{
  "6": 0.49322128674835697
}
{
  "7": 0.1878849191181694
}
{}
{
  "9": 0.7495507603507059
}
{
  "10": 0.9526183602969864
}

mlr --from ./reg-test/input/abixy put 
  func f(int x): map {
    if (NR==2) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($x)


mlr --from ./reg-test/input/abixy put 
  func f(int x): map {
    if (NR==200) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($i)

a=pan,b=pan,i=1,x=0.3467901443380824,y={}
a=eks,b=pan,i=2,x=0.7586799647899636,y={}
a=wye,b=wye,i=3,x=0.20460330576630303,y={}
a=eks,b=wye,i=4,x=0.38139939387114097,y={}
a=wye,b=pan,i=5,x=0.5732889198020006,y={}
a=zee,b=pan,i=6,x=0.5271261600918548,y={}
a=eks,b=zee,i=7,x=0.6117840605678454,y={}
a=zee,b=wye,i=8,x=0.5985540091064224,y={}
a=hat,b=wye,i=9,x=0.03144187646093577,y={}
a=pan,b=wye,i=10,x=0.5026260055412137,y={}

mlr --from ./reg-test/input/abixy put 
  func f(int x): map {
    if (NR==200) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($x)


mlr --from ./reg-test/input/abixy put 
  func f(int x): var {
    if (NR==2) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($i)

a=pan,b=pan,i=1,x=0.3467901443380824,y={}
a=eks,b=pan,i=2,x=0.7586799647899636,y=2
a=wye,b=wye,i=3,x=0.20460330576630303,y={}
a=eks,b=wye,i=4,x=0.38139939387114097,y={}
a=wye,b=pan,i=5,x=0.5732889198020006,y={}
a=zee,b=pan,i=6,x=0.5271261600918548,y={}
a=eks,b=zee,i=7,x=0.6117840605678454,y={}
a=zee,b=wye,i=8,x=0.5985540091064224,y={}
a=hat,b=wye,i=9,x=0.03144187646093577,y={}
a=pan,b=wye,i=10,x=0.5026260055412137,y={}

mlr --from ./reg-test/input/abixy put 
  func f(int x): var {
    if (NR==2) {
      return 2
    } else {
      return {}
    }
  }
  $y=f($x)


mlr --from ./reg-test/input/abixy put 
  func f(x): int {
    # fall-through return value is absent-null
  }
  $y=f($x)


mlr --from ./reg-test/input/abixy put 
  int a = 1;
  var b = a[2]; # cannot index localvar declared non-map


mlr --oxtab --from ./reg-test/input/abixy-het put 
 $* = mapdiff(
   mapsum($*, {"a": "newval"}),
   {"b": "nonesuch"},
 )

a newval
i 1
x 0.3467901443380824
y 0.7268028627434533

a newval
i 2
x 0.7586799647899636
y 0.5221511083334797

aaa wye
i   3
x   0.20460330576630303
y   0.33831852551664776
a   newval

a   newval
bbb wye
i   4
x   0.38139939387114097
y   0.13418874328430463

a   newval
i   5
xxx 0.5732889198020006
y   0.8636244699032729

a newval
i 6
x 0.5271261600918548
y 0.49322128674835697

a   newval
iii 7
x   0.6117840605678454
y   0.1878849191181694

a   newval
i   8
x   0.5985540091064224
yyy 0.976181385699006

aaa hat
bbb wye
i   9
x   0.03144187646093577
y   0.7495507603507059
a   newval

a newval
i 10
x 0.5026260055412137
y 0.9526183602969864

